### PR TITLE
[fbrp] autocomplete proc names

### PR DESCRIPTION
**Pull request by lshamis**
# Description

Add auto-completion to process names, in addition to commands.

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [x] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

Before
`fbrp u<TAB>`  -> `fbrp up`
`fbrp up <TAB>`  -> `fbrp up`

After
`fbrp u<TAB>`  -> `fbrp up`
`fbrp up <TAB>`  -> `fbrp up alice`

# Testing

Added new unit tests.
Manually tested all commands across several examples.

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [x] I have added tests that show that the PR is functional.
- [x] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

### CODE REVIEW

File: fbrp/src/fbrp/cmd/logs.py [hunk 1]
```diff
@@ -8,11 +9,22 @@
 
 
 @click.command()
-@click.argument("procs", nargs=-1)
+@click.argument(
+    "procs",
+    nargs=-1,
+    shell_complete=_autocomplete.conditional(
```
Comment (1heart) [inline fbrp/src/fbrp/cmd/logs.py:15]: Add a comment explaining the logic here -- requires some time thinking about the higher-order function to parse this.

**Review by 1heart — APPROVED** [review]
Cool, thanks! If there's a way to simplify the higher-order functional logic, might be good for readability. Maybe naming all functions in the same way when they follow the click interface (`lambda ctx, param, incomplete: list_of_strs`) , then the higher-order functions separately.

File: fbrp/src/fbrp/cmd/_autocomplete.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+def alephzero_topics(protocol):
+    def fn(ctx, param, incomplete):
+        topics = []
+        detected = glob.glob(
+            os.path.join(a0.env.root(), f"**/*.{protocol}.a0"), recursive=True
+        )
+        for abspath in detected:
+            relpath = os.path.relpath(abspath, a0.env.root())
+            topic = relpath[: -len(f".{protocol}.a0")]
+            topics.append(topic)
```
[the remark this answers is not in the document]
Comment (lshamis) [reply fbrp/src/fbrp/cmd/_autocomplete.py:68]: Added a TODO

File: fbrp/src/fbrp/cmd/logs.py (hunk 1 above)
Comment (lshamis) [reply fbrp/src/fbrp/cmd/logs.py:19]: Added a comment

**Comment by lshamis** [comment]
@1heart I might be able to simplify the notation by making the functions into classes with a `__call__(ctx, param, incomplete)`. Then I can add operators such as `|`, `&`, ...
I'm not sure how to make conditionals cleaner.

I think I'll tackle this in a followup.

**Pull request merged**